src/uiScreens.c

```diff
@@ -173,31 +173,33 @@ void ui_displayStakingInfoScreen(
 	);
 }
 
-void ui_displayTokenNameScreen(
+#define ASSET_FINGERPRINT_SIZE 20
+
+void ui_displayAssetFingerprintScreen(
+        token_group_t* tokenGroup,
         token_amount_t* token,
         ui_callback_fn_t callback
 )
 {
-	if (str_isAsciiPrintableBuffer(token->assetNameBytes, token->assetNameSize)) {
-		char name[ASSET_NAME_SIZE_MAX + 1];
-		ASSERT(token->assetNameSize + 1 <= SIZEOF(name));
-		os_memmove(name, token->assetNameBytes, token->assetNameSize);
-		name[token->assetNameSize] = '\0';
+	uint8_t hashInput[MINTING_POLICY_ID_SIZE + ASSET_NAME_SIZE_MAX];
+	const size_t hashInputSize = MINTING_POLICY_ID_SIZE + token->assetNameSize;
+	{
+		write_view_t view = make_write_view(hashInput, hashInput + SIZEOF(hashInput));
+		ASSERT(SIZEOF(tokenGroup->policyId) == MINTING_POLICY_ID_SIZE);
+		view_appendData(&view, tokenGroup->policyId, MINTING_POLICY_ID_SIZE);
+		view_appendData(&view, token->assetNameBytes, token->assetNameSize);
+		ASSERT(view_processedSize(&view) == hashInputSize);
+	}
 
-		bool isEmpty = (token->assetNameSize == 0);
+	uint8_t fingerprint[ASSET_FINGERPRINT_SIZE];
+	blake2b_160_hash(hashInput, hashInputSize, fingerprint, SIZEOF(fingerprint));
 
-		ui_displayPaginatedText(
-		        (isEmpty) ? "Asset name is empty" : "Asset name (ASCII):",
-		        name,
-		        callback
-		);
-	} else {
-		ui_displayHexBufferScreen(
-		        "Asset name (hex):",
-		        token->assetNameBytes, token->assetNameSize,
-		        callback
-		);
-	}
+	ui_displayBech32Screen(
```

Comment:
I'd rather not couple bech32 encoding with screen rendering logic. Also, I don't see this function being reused for other contexts where it would make sense like showing pool id or vrf key in the pool registration certificate

Reply:
It is reused in pool operator app, I actually extracted it from there.

Reply:
This function (similarly to hexBufferScreen etc.) provides a crucial feature that is not obvious: it allocates the buffer for the bech32 string locally. If I have a ui handler with, say, 5 steps, and each of them needs a string buffer, I cannot afford to allocate all of the buffers in the handler because there is not enough stack space. So a separate function, marked __noinline_due_to_stack__, must be called (see the operator app PR). And it is exactly what this function does: encodes a buffer into bech32 and displays it. I don't see any way around this if we are only using static memory.